### PR TITLE
[vividus] Use Service Loader for gathering Properties Processors

### PR DESCRIPTION
**Pull request by abudevich**

### CODE REVIEW

File: vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java [hunk 1]
```diff
@@ -144,14 +145,10 @@ public static ConfigurationResolver getInstance() throws IOException
         deprecatedPropertiesHandler.removeDeprecated(properties);
         resolveSpelExpressions(properties, false);
 
-        try (VaultStoredPropertiesProcessor vaultProcessor = new VaultStoredPropertiesProcessor(properties))
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java:147]: we still need to close `VaultStoredPropertiesProcessor`

File: vividus/src/main/java/org/vividus/configuration/DelegatingPropertiesProcessor.java [hunk 2]
```diff
@@ -19,13 +19,15 @@
...
+
 public class DelegatingPropertiesProcessor implements PropertiesProcessor
 {
     private final List<PropertiesProcessor> propertiesProcessors;
 
-    public DelegatingPropertiesProcessor(List<PropertiesProcessor> propertiesProcessors)
+    public DelegatingPropertiesProcessor(Iterable<PropertiesProcessor> propertiesProcessors)
     {
-        this.propertiesProcessors = propertiesProcessors;
+        this.propertiesProcessors = Lists.newArrayList(propertiesProcessors);
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/configuration/DelegatingPropertiesProcessor.java:30]: no reason to convert to list, `Iterable` can be used

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-engine/src/main/java/org/vividus/configuration/PropertiesProcessor.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2023 the original author or authors.
+ * Copyright 2019-2024 the original author or authors.
```
[the remark this answers is not in the document]
Comment (abudevich) [reply vividus-engine/src/main/java/org/vividus/configuration/PropertiesProcessor.java:2]: due to moving this file I have exception during build:
`Execution failed for task ':vividus-engine:spotlessJavaCheck'.`

File: vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java [hunk 4]
```diff
@@ -144,14 +145,23 @@ public static ConfigurationResolver getInstance() throws IOException
...
-            ));
-            new SystemPropertiesInitializer(propertiesProcessor).setSystemProperties(properties);
-            properties = propertiesProcessor.processProperties(properties);
-        }
+        ServiceLoader<PropertiesProcessor> loader = ServiceLoader.load(PropertiesProcessor.class);
+        PropertiesProcessor propertiesProcessor = new DelegatingPropertiesProcessor(loader);
+        properties = propertiesProcessor.processProperties(properties);
+        new SystemPropertiesInitializer(propertiesProcessor).setSystemProperties(properties);
+        loader.forEach(processor -> {
+            if (processor instanceof VaultStoredPropertiesProcessor)
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/configuration/ConfigurationResolver.java:153]: `ConfigurationResolver` should know nothing about particular implementations of `PropertiesProcessor`, it should be designed in this way:
- `PropertiesProcessor` should inherit `Closeable` interface
- `PropertiesProcessor` should have default empty implementation of `close()`
- `DelegatingPropertiesProcessor` in its `close` method should invoke `close` for each delegate
- `DelegatingPropertiesProcessor` instance should be used with `try-with-resources`

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — COMMENTED** [review]
Since the approach is changed, `processProperty` method should be moved from `PropertiesProcessor` interface to `AbstractPropertiesProcessor` class

File: vividus/src/main/java/org/vividus/configuration/DelegatingPropertiesProcessor.java [hunk 5]
```diff
@@ -16,29 +16,45 @@
...
+
+    @Override
+    public void close()
+    {
+        propertiesProcessors.iterator().forEachRemaining(p -> {
+            try
+            {
+                p.close();
+            }
+            catch (IOException e)
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/configuration/DelegatingPropertiesProcessor.java:54]: just throw `IOException`

File: vividus/src/main/java/org/vividus/configuration/VaultStoredPropertiesProcessor.java [hunk 6]
```diff
@@ -39,14 +39,21 @@ public class VaultStoredPropertiesProcessor extends AbstractPropertiesProcessor
 {
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/configuration/VaultStoredPropertiesProcessor.java:39]: it should be updated to remove `implements Closeable`

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request merged**